### PR TITLE
#1119 added new TemplatesUtilsClassesUtil class to core (needs to be …

### PR DESCRIPTION
**Pull request by jan-vcapgemini**
…used by eclipse-plugin, maven-plugin and cobigen-cli)

Adresses/Fixes #1119 #1110 .

Implements

* new centralized ResolveUtilClasses handling
* 
* 

@devonfw/cobigen

### CODE REVIEW

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+        } else {
+            LOG.info("Processing configuration folder " + templateRoot.toString());
+            LOG.debug("Searching for classes ...");
+            final List<Path> foundPaths = new LinkedList<>();
+
+            try {
+                Files.walkFileTree(templateRoot, new SimpleFileVisitor<Path>() {
+                    @Override
+                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                        if (file.toString().endsWith(".class")) {
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java:205]: I think path api also provides an getExtension api so you don't need tostring resulting in bad performance

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+     * @throws IOException
+     *             if either templates jar or templates folder could not be read
+     */
+    public List<Class<?>> resolveUtilClasses(File configurationFolder) throws IOException {
+        final List<Class<?>> result = new LinkedList<>();
+
+        Path templateRoot = null;
+        ClassLoader inputClassLoader = null;
+        if (configurationFolder != null) {
+            addUrlToClassLoaderUrls(configurationFolder.toURI().toURL());
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java:129]: Yes, I'm not happy with the getContextConfiguration method yet, currently it requires the ClassLoader to include the whole configuration folder path to detect the context.xml properly.
I'm open for suggestions.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+import java.util.Map;
+import java.util.Queue;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ */
+public class TemplatesUtilsClassesUtil {
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: I've renamed the class and added a javadoc now.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+    /** Logger instance. */
+    private static final Logger LOG = LoggerFactory.getLogger(TemplatesJarUtil.class);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: fixed.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+    /**
+     * Adds the given URL to the classLoaderUrls array
+     * @param url
+     *            URL to add to classLoaderUrls array
+     */
+    private void addUrlToClassLoaderUrls(URL url) {
+        ArrayList<URL> newUrls = new ArrayList<>(Arrays.asList((classLoaderUrls)));
+        newUrls.add(url);
+        classLoaderUrls = newUrls.toArray(new URL[] {});
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: I've replaced the whole method with an ArrayList.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+                        LOG.info("    * Removed class file " + next.toString());
+                        it.remove();
+                    }
+                }
+
+                for (Path path : foundPaths) {
+                    try {
+                        result.add(loadClassByPath(templateRoot.relativize(path), inputClassLoader));
+                    } catch (ClassNotFoundException e) {
+                        LOG.error("Class could not be loaded into ClassLoader " + e);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: fixed.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+     * @param configurationFolder
+     *            File configuration folder for which to generate paths
+     * @throws MalformedURLException
+     *             if the URL was malformed
+     */
+    private void addFoldersToClassLoaderUrls(File configurationFolder) throws MalformedURLException {
+        String[] possibleLocations = new String[] { "src/main/templates", "target/classes" };
+
+        for (String possibleLocation : possibleLocations) {
+            File folder = Paths.get(configurationFolder + File.separator + possibleLocation).toFile();
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+            Paths.get(uri);
+
+            try {
+                // Get the URI of the jar from the URL of the contained context.xml
+                URI jarUri = URI.create(contextConfigurationLocation.toString().split("!")[0]);
+                FileSystem jarfs = FileSystems.getFileSystem(jarUri);
+
+                // walk the jar file
+                LOG.debug("Searching for classes in " + jarUri.toString());
+                Files.walkFileTree(jarfs.getPath("/"), new SimpleFileVisitor<Path>() {
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: Moved to 2 new methods, one for the folder and one for the jar.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+                    }
+                });
+            } catch (IOException e) {
+                LOG.error("Could not read templates jar file " + e);
+            }
+            for (String className : foundClasses) {
+                try {
+                    result.add(inputClassLoader.loadClass(className));
+                } catch (ClassNotFoundException e) {
+                    LOG.warn("Could not load " + className + " from classpath", e);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+
+                    @Override
+                    public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                        // Log errors but do not throw an exception
+                        LOG.warn("visitFileFailed @ " + exc);
+                        return FileVisitResult.CONTINUE;
+                    }
+                });
+            } catch (IOException e) {
+                LOG.error("Could not read templates jar file " + e);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+                            // list of found files
+                            foundClasses.add(fileName.replace("/", "."));
+                        }
+                        return FileVisitResult.CONTINUE;
+                    }
+
+                    @Override
+                    public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                        // Log errors but do not throw an exception
+                        LOG.warn("visitFileFailed @ " + exc);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+
+                    @Override
+                    public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                        // Log errors but do not throw an exception
+                        LOG.warn("visitFileFailed @" + exc);
+                        return FileVisitResult.CONTINUE;
+                    }
+                });
+            } catch (IOException e) {
+                LOG.error("Could not read templates folder " + e);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+                            foundPaths.add(file);
+                            LOG.debug("    * Found class file " + file.toString());
+                        }
+                        return FileVisitResult.CONTINUE;
+                    }
+
+                    @Override
+                    public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                        // Log errors but do not throw an exception
+                        LOG.warn("visitFileFailed @" + exc);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+ */
+public class TemplatesUtilsClassesUtil {
+
+    /** Logger instance. */
+    private static final Logger LOG = LoggerFactory.getLogger(TemplatesJarUtil.class);
+
+    /**
+     * Stores the URLs for the ClassLoader
+     */
+    private URL[] classLoaderUrls = {};
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: Moved URL ArrayList into respective methods.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java (hunk 1 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+     * templates-plugin and returns its URL
+     * @param classLoader
+     *            ClassLoader to check resources from
+     * @return URL of the context configuration file path
+     * @throws IOException
+     *             if no configuration file was found
+     */
+    public URL getContextConfiguration(ClassLoader classLoader) throws IOException {
+        URL contextConfigurationLocation = null;
+        String[] possibleLocations = new String[] { "context.xml", "src/main/templates/context.xml" };
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:48]: This static array is created again and again. Better make it a static final field to handle it properly in memory and allow the reader to understand that this array will never be changed

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+            URL configLocation = classLoader.getResource(possibleLocation);
+            if (configLocation != null) {
+                contextConfigurationLocation = configLocation;
+                LOG.debug("Found context.xml URL in the classpath @ {}", contextConfigurationLocation);
+                break;
+            }
+        }
+
+        if (contextConfigurationLocation == null) {
+            throw new IOException("No context.xml could be found in the classpath!");
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:60]: Better create a configuration exception which is already available in cobigen which indicates a problem with the configuration and is properly handled by eclipse or maven plugin already

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+    /**
+     * Adds folders to class loader urls e.g. src/main/templates for config.xml detection
+     * @param configurationFolder
+     *            Path configuration folder for which to generate paths
+     * @return ArrayList<URL> of URLs
+     * @throws MalformedURLException
+     *             if the URL was malformed
+     */
+    private ArrayList<URL> addFoldersToClassLoaderUrls(Path configurationFolder) throws MalformedURLException {
+        String[] possibleLocations = new String[] { "src/main/templates", "target/classes" };
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:86]: Same here. Better a private static final field.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+     * @return Class<?> of the class file
+     * @throws ClassNotFoundException
+     *             if no class could be found all the way up to the path root
+     */
+    public Class<?> loadClassByPath(Path classPath, ClassLoader cl) throws ClassNotFoundException {
+        // Get a list with all path segments, starting with the class name
+        Queue<String> pathSegments = new LinkedList<>();
+        // Split the path by the systems file separator and without the .class suffix
+        String[] pathSegmentsArray = classPath.toString().substring(0, classPath.toString().length() - 6)
+            .split("\\".equals(File.separator) ? "\\\\" : File.separator);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 15 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 17 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+            env.put("create", "true");
+
+            URI uri = URI.create(contextConfigurationLocation.toString());
+            FileSystem fs;
+            try {
+                fs = FileSystems.getFileSystem(uri);
+            } catch (FileSystemNotFoundException e) {
+                fs = FileSystems.newFileSystem(uri, env);
+            }
+            Paths.get(uri);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesUtilsClassesUtil.java]: Adjusted to getOrCreateFileSystem from core.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 16 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: Adjusted to InvalidConfigurationException.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 20]
```diff
@@ -0,0 +1,351 @@
...
+            foundClasses = walkJarFile(jarUri, jarfs);
+        } catch (IOException e) {
+            LOG.error("Could not read templates jar file", e);
+        }
+        if (foundClasses.size() > 0) {
+            for (String className : foundClasses) {
+                try {
+                    result.add(inputClassLoader.loadClass(className));
+                } catch (ClassNotFoundException e) {
+                    LOG.warn("Could not load " + className + " from classpath");
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:223]: Remember correct string template syntax

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 21]
```diff
@@ -0,0 +1,351 @@
...
+                return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+                // Log errors but do not throw an exception
+                LOG.warn("An IOException occurred while reading file on path {} with message: {}", file,
+                    exc.getMessage());
+                LOG.debug("An IOException occurred while reading file on path {} with message: {}", file,
+                    LOG.isDebugEnabled() ? exc : null);
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:303]: You can put this last parameter simply in the warn log before

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:260]: You can put this last parameter simply in the warn log before

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 23]
```diff
@@ -0,0 +1,351 @@
...
+            LOG.error("Could not read templates folder", e);
+        }
+        if (foundPaths.size() > 0) {
+
+            // clean up test classes
+            Iterator<Path> it = foundPaths.iterator();
+            while (it.hasNext()) {
+                Path next = it.next();
+                if (!templateRoot.relativize(next).startsWith("target/classes")) {
+                    LOG.info("    * Removed test class file {}", next);
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:171]: Better on debug as there might be a lot test classes

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+     *            List to store utility classes in
+     * @param templateRoot
+     *            Path to template folder containing classes
+     * @param inputClassLoader
+     *            ClassLoader to use for storing of classes
+     * @return List of classes to load utilities from
+     */
+    private static List<Class<?>> resolveFromFolder(List<Class<?>> result, Path templateRoot,
+        ClassLoader inputClassLoader) {
+        LOG.debug("Processing configuration folder " + templateRoot.toString());
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:155]: Remember string template syntax

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 22 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:260]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 24 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 23 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 21 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:303]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 20 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 25]
```diff
@@ -255,9 +255,9 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
             public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
                 // Log errors but do not throw an exception
                 LOG.warn("An IOException occurred while reading file on path {} with message: {}", file,
-                    exc.getMessage());
+                    exc.getMessage(), LOG.isDebugEnabled() ? exc : null);
                 LOG.debug("An IOException occurred while reading file on path {} with message: {}", file,
-                    LOG.isDebugEnabled() ? exc : null);
+                    exc.getMessage(), LOG.isDebugEnabled() ? exc : null);
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:260]: remove duplicated debug log as this is already covered by warn

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java [hunk 26]
```diff
@@ -298,9 +298,9 @@ public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IO
...
```
Comment (maybeec) [inline cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java:303]: remove duplicated debug log as this is already covered by warn

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 25 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: adjusted.

File: cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java (hunk 26 above)
Comment (jan-vcapgemini) [reply cobigen/cobigen-core-parent/cobigen-core/src/main/java/com/devonfw/cobigen/impl/util/TemplatesClassloaderUtil.java]: adjusted.

**Pull request merged**